### PR TITLE
10587 script pagination

### PR DESCRIPTION
**Pull request by arthanson**
### Fixes: #10587 

Makes the script logs use a tables2 table which allows pagination and sorting of columns.  There were a couple bugs in the scripts migrations which this fixes (was causing jobs not to get correctly mapped from old script/report runs).

Original FR asked for filtering and export, however the logs are stored as a dict and filtering requires a queryset so filtering is not possible without converting the individual logs lines to be stored as regular DB objects.
Table configuration is implemented but not tested as table configuration is currently broken across all views.

Monosnap DeviceConnectionsReport | NetBox 2024-03-04 13-44-08

### CODE REVIEW

File: netbox/extras/tables/tables.py [hunk 1]
```diff
@@ -507,3 +509,67 @@ class Meta(NetBoxTable.Meta):
...
+    )
+
+    class Meta(BaseTable.Meta):
+        empty_text = _('No results found')
+        fields = (
+            'index', 'method', 'time', 'status', 'object', 'url', 'message',
+        )
+        default_columns = (
+            'index', 'method', 'time', 'status', 'object', 'url', 'message',
+        )
```
Comment (jeremystretch) [inline netbox/extras/tables/tables.py:575]:
```suggestion
```

File: netbox/extras/tables/tables.py [hunk 2]
```diff
@@ -507,3 +509,67 @@ class Meta(NetBoxTable.Meta):
...
+    status = tables.TemplateColumn(
+        template_code="""{% load log_levels %}{% log_level record.status %}""",
+        verbose_name=_('Level')
+    )
+
+    object = tables.Column(
+        verbose_name=_('Object')
+    )
+    url = tables.Column(
+        verbose_name=_('Url')
```
Comment (jeremystretch) [inline netbox/extras/tables/tables.py:562]:
```suggestion
        verbose_name=_('URL')
```

File: netbox/extras/views.py [hunk 3]
```diff
@@ -1143,19 +1145,71 @@ def get(self, request, module, name, path=''):
...
+                        data.append(result)
+
+            table = ReportResultsTable(data, user=request.user)
+            table.configure(request)
+
+        return table
+
     def get(self, request, **kwargs):
         job = get_object_or_404(Job.objects.all(), pk=kwargs.get('job_pk'))
+        table_logs = table_tests = None
```
Comment (jeremystretch) [inline netbox/extras/views.py:1202]: Unused?
```suggestion
```

File: netbox/extras/views.py [hunk 4]
```diff
@@ -1143,19 +1145,71 @@ def get(self, request, module, name, path=''):
...
+
+        if job.completed:
+            table = self.get_table(job, request, bulk_actions=False)
 
         context = {
             'script': job.object,
             'job': job,
+            'table': table,
```
Comment (jeremystretch) [inline netbox/extras/views.py:1210]: `table` will be undefined if `job.completed` is False.

File: netbox/extras/views.py [hunk 5]
```diff
@@ -1143,19 +1145,71 @@ def get(self, request, module, name, path=''):
...
+        tests = None
+        table = None
+        index = 0
+        if job.data:
+            if 'log' in job.data:
+                if 'tests' in job.data:
+                    tests = job.data['tests']
+
+                for log in job.data['log']:
+                    index += 1
```
Comment (jeremystretch) [inline netbox/extras/views.py:1165]: We can avoid having to track the index manually by wrapping `job.data['log']` with `enumerate()`.

File: netbox/extras/views.py [hunk 6]
```diff
@@ -1143,19 +1145,71 @@ def get(self, request, module, name, path=''):
...
+                    result = {
+                        'index': index,
+                        'time': log.get('time', None),
+                        'status': log.get('status', None),
+                        'message': log.get('message', None),
```
Comment (jeremystretch) [inline netbox/extras/views.py:1170]: Specifying `None` as the default is redundant if `log` is a dict.
```suggestion
                        'time': log.get('time'),
                        'status': log.get('status'),
                        'message': log.get('message'),
```

File: netbox/extras/views.py [hunk 7]
```diff
@@ -1143,19 +1145,71 @@ def get(self, request, module, name, path=''):
...
+                    }
+                    data.append(result)
+
+                table = ScriptResultsTable(data, user=request.user)
+                table.configure(request)
+            else:
+                tests = job.data
```
Comment (jeremystretch) [inline netbox/extras/views.py:1177]: This should have a comment explaining the accommodation for legacy reports.

File: netbox/extras/views.py [hunk 8]
```diff
@@ -1143,19 +1145,71 @@ def get(self, request, module, name, path=''):
...
+
+        if tests:
+            for method, test_data in tests.items():
+                if 'log' in test_data:
+                    for time, status, obj, url, message in test_data['log']:
+                        index += 1
```
Comment (jeremystretch) [inline netbox/extras/views.py:1183]: Same as above; we can call `enumerate()` and pass it the previous index value (if any) as the starting value.

**Review by jeremystretch — CHANGES_REQUESTED** [review]

File: netbox/extras/views.py (hunk 8 above)
Comment (arthanson) [reply netbox/extras/views.py:1184]: I didn't do this one as this second one potentially relies on the count of the first one, so you would need a start_counter defined at top then conditionally set the enumeration start to either the start_counter or the end of the previous list, so I thought index was just more readable as to what is going on.

File: netbox/extras/views.py (hunk 5 above)
Comment (arthanson) [reply netbox/extras/views.py:1165]: see comment below

**Review by jeremystretch — APPROVED** [review]

**Pull request merged**